README.md
# pe_network-cookbook

Cookbook to control network devices, physical mappings, internet addresses and firewall 
rules on a server. It can read all the attributes from a databag.

Warning, For the moment, this cookbook does not control the GW!

## Supported Platforms

 * Debian
 * Ubuntu
 * Centos

This cookbook does not work in Debian platforms with chef version 11.4

## Attributes

To can define the attributes, or use a databag to read and setup all of them.
For instance, here you see the attribute file of this cookbook:

```
### Main databag attributes
default[:pe_network][:data_bag] = 'network'
default[:pe_network][:bag_name] = node[:hostname]
default[:pe_network][:environment] = node[:chef_environment] ? node[:chef_environment] : "_default"

### Network Definitions
default[:pe_network][:udev] = "/etc/udev/rules.d/70-persistent-net.rules"
default[:pe_network][:network] = {
#
# Linking an IP to a device:
#  'eth0' => {
#      :ip => "192.168.1.2",
#      :network => "192.168.1.0",
#      :mask => "255.255.255.0",
#      :mac => "08:00:27:84:ee:e7",
#      :mtu => "1400",
#      :bcast => "192.168.1.255",
#      :onboot => "yes"
#  },
#
# Forcing the name of a device by the MAC:
#  'eth1' => {
#      :mac => "08:00:27:84:ee:e7"
#  }
}
# Do not run the FW recipe by default
default[:pe_network][:run_fw] = false

### Firewall
default[:pe_network][:firewall][:enabled] = false
default[:pe_network][:firewall][:log] = true
# DROP all input packets by defaut
default[:pe_network][:firewall][:enable_input_drop] = true
# DROP all output packets by defaut
default[:pe_network][:firewall][:enable_output_drop] = false
# Rules
default[:pe_network][:firewall][:rules] = {
   'allow ssh' => {
      :direction => 'in',
      :user => 'root',
      :protocol => 'tcp',
      :source => ['0.0.0.0/0'],
      :dport => '22'
    }
#  '<rule name>' =>
#     :direction => '<in|out>',
#     :protocol => '<udp|tcp|icmp>',
#     :user => '<local user from /etc/passwd>',
#     :interface => '<default|all|eth0|eth1|br0|...>',
#     :source => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>',
#     :sport => '<integer(:integer))>',
#     :destination => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>',
#     :dport => '<integer(:integer)>',
#     :env => '<production|staging|...>',
#     :options => ['disable_env_limit', 'disable_syntax_check', ...]
}
```
All the default value of the attributes are safe and the best way to define the 
parameters is using a databag.


## Usage

The easy way is just use a databag to define the device mappings, ips and firewall rules. 
Be carefull defining the rules on the primary interface ... you can lost the connectivity 
in the middle of the chef convergence. So, you have to create a databag named `network` 
with a file named equal as the node's name. In that file you can define all the attributes 
in json format, for example:

```json
{
    "id": "pe-network-berkshelf",
    "_default": {
        "network": {
           "eth1": {
               "ip": "192.168.1.2",
               "mac": "08:00:27:84:ee:e7",
               "mtu": "1400"
           }
        },
        "firewall": {
            "enabled": true,
            "enable_input_drop": true,
            "enable_output_drop": false,
            "rules": {
                "allow ssh": {
                    "direction": "in",
                    "user": "root",
                    "protocol": "tcp",
                    "source": ["0.0.0.0/0"],
                    "dport": "22"
                },
                "allow web": {
                    "direction": "in",
                    "user": "root",
                    "protocol": "tcp",
                    "source": ["0.0.0.0/0"],
                    "dport": "80"
                }
        }
    }
}
```

The configuration is pretty easy to understand. There are two parts, the first part `network` 
is to define the ifconfig parameters (it also supports Centos and RH):

 * `device` is the only mandatory attribute, the cookbook will assign the IP to it.
 * If there is a `mac`, it will change or create the udev rules to assure the mapping (even if the mac is not found).
 * You can define other attributes like:
     * `mtu`
     * `network`
     * `mask`
     * `bcast` 

In the second part you can define all the firewall rules. You can use chef searchs in the source and destination attributes :

 * 'rule name' =>
     * `direction` => '<in|out>'  (*mandatory*)
     * `protocol` => '<udp|tcp|icmp>' (*mandatory*)
     * `user` => '<local user from /etc/passwd>' (*mandatory*)
     * `interface` => '<default|all|eth0|eth1|br0|...>'
     * `source` => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>' (*mandatory for `in` rules*)
     * `sport` => '<integer(:integer))>'
     * `destination` => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>' (*mandatory for `out` rules*)
     * `dport` => '<integer(:integer)>'
     * `env` => '<production|staging|...>',
     * `options` => ['disable_env_limit', 'disable_syntax_check', ...]

You can use chef searchs in the source and destination attributes. More info in the
AFW's home site: https://github.com/jvehent/AFW

To apply the cookbook just include `pe_network` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[pe_network::default]"
  ]
}
```

## License and Authors

Author:: Jose Riguera Lopez, Springer SBM (<jose.riguera@springer.com>)


CHANGELOG.md
# 0.1.2

Added :run_fw = false attribute to control the FW recipe

# 0.1.1

Changed attribute hash to the device the key, so it is possible to define device <-> mac link without ip 

# 0.1.0

Initial release of pe_network

* Support to update and create udev rules for devices depending on the mac address
* Link inet addr to a device
* Setup firewall rules


recipes/network.rb
#
# Cookbook Name:: pe_network
# Recipe:: network
#
# Copyright (C) 2014 Jose Riguera, Springer SBM
# 

class ::Chef::Recipe
  include SPRpe
end
Chef::Resource::RubyBlock.send(:include, SPRpe)

udev = node[:pe_network][:udev]
networks = node[:pe_network][:network]
begin
   networks.each_pair do |device, parameters|
      # Process the settings
      if parameters[:mac]
         mac = parameters[:mac]
         dev = dev_addr(mac)
         Chef::Log.info("Current device for #{mac}: #{dev}")
         if !(device && device == dev)
            device = device ? device : dev            
            ruby_block "update_udev" do
               block do
                  net_udev(mac, device, true, udev)
               end
            end
         end
      end
      ip = parameters[:ip] ? parameters[:ip] : nil
      if ip
         Chef::Log.info("Setting '#{ip}' on '#{device}'")
         # Set the IP for each device
         ifconfig device do
            ignore_failure true
  	    device device
            action :delete
            notifies :add, "ifconfig[#{ip}]", :immediately
         end
         ifconfig ip do
            ignore_failure false
  	    device device
            target ip
            network parameters[:network] ? parameters[:network] : nil
            mtu parameters[:mtu] ? parameters[:mtu] : nil
            mask parameters[:mask] ? parameters[:mask] : nil
            bcast parameters[:bcast] ? parameters[:bcast] : nil
            onboot parameters[:onboot] ? parameters[:onboot] : nil
            action :add
            #notifies :restart, "service[networking]", :delayed
         end
      end
   end
rescue
   Chef::Log.fatal("Processing network")
end

#service "networking" do
#   service_name "networking"
#   action :restart
#end



recipes/firewall.rb
#
# Cookbook Name:: pe_network
# Recipe:: firewall
#
# Copyright (C) 2014 Jose Riguera, Springer SBM
# 

class ::Chef::Recipe
  include SPRpe
  include AFW
end

# enable or disable the firewall restore command. If set the false, 
# the rules will still be populated in /etc/firewall/rules.iptables 
# but the restore command will not be issued.
node.override[:afw][:enable] = node[:pe_network][:firewall][:enabled]
node.override[:afw][:enable_input_drop] = node[:pe_network][:firewall][:enable_input_drop]
node.override[:afw][:enable_output_drop] = node[:pe_network][:firewall][:enable_output_drop]
node.override[:afw][:enable_input_drop_log] = node[:pe_network][:firewall][:log]
node.override[:afw][:enable_output_drop_log] = node[:pe_network][:firewall][:log]

# Include the recipe to create the default configuration
include_recipe 'afw::default'

fw = node[:pe_network][:firewall][:rules]
begin
   fw.each_pair do |rule, parameters|
      # Process the settings
      # Call the AFW module to create the rule
      AFW.create_rule(node, rule, parameters)
   end
rescue
   Chef::Log.fatal("Processing firewall rules")
end



attributes/default.rb
# pe_network

### Main attributes
default[:pe_network][:data_bag] = 'network'
default[:pe_network][:bag_name] = node[:hostname]
default[:pe_network][:environment] = node[:chef_environment] ? node[:chef_environment] : "_default"

### Network Definitions
default[:pe_network][:udev] = "/etc/udev/rules.d/70-persistent-net.rules"
default[:pe_network][:network] = {
#  'eth0' => {
#      :ip => "192.168.1.2",
#      :network => "192.168.1.0",
#      :mask => "255.255.255.0",
#      :mac => "08:00:27:84:ee:e7",
#      :mtu => "1400",
#      :bcast => "192.168.1.255",
#      :onboot => "yes"
#  },
#  'eth1' => {
#      :mac => "08:00:27:84:ee:e7"
#  }
}
# Do not run the FW recipe by default
default[:pe_network][:run_fw] = false

### Firewall
default[:pe_network][:firewall][:enabled] = false
default[:pe_network][:firewall][:log] = false
# DROP all input packets by defaut
default[:pe_network][:firewall][:enable_input_drop] = true
# DROP all output packets by defaut
default[:pe_network][:firewall][:enable_output_drop] = false
# Rules
default[:pe_network][:firewall][:rules] = {
   'allow ssh' => {
      :direction => 'in',
      :user => 'root',
      :protocol => 'tcp',
      :source => ['0.0.0.0/0'],
      :dport => '22'
    }
#  '<rule name>' =>
#     :direction => '<in|out>',
#     :protocol => '<udp|tcp|icmp>',
#     :user => '<local user from /etc/passwd>',
#     :interface => '<default|all|eth0|eth1|br0|...>',
#     :source => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>',
#     :sport => '<integer(:integer))>',
#     :destination => '<ip|fqdn|chef search>|['<ip|fqdn|chef search>',...]'>',
#     :dport => '<integer(:integer)>',
#     :env => '<production|staging|...>',
#     :options => ['disable_env_limit', 'disable_syntax_check', ...]
}

